### ISSUE TITLE
Bug in hipsycl::sycl::exception

### ISSUE THREAD

**Issue by breyerml** [AdaptiveCpp/AdaptiveCpp#603]
**Bug summary**
Returning a `const&` to a temporary object is UB.

**To Reproduce**
Compiling a simple test program 
```cpp
#include <iostream>
#include <sycl/sycl.hpp>

using namespace sycl;

int main() {
  constexpr std::size_t size = 64;
  
  queue myQueue;

  int *data = sycl::malloc_device<int>(size, myQueue);
  
  event e = myQueue.parallel_for(range<1>{ size }, [=](id<1> idx) {
    data[idx.get(0)] = idx.get(0);
  });
  
  int hostData[size];
  myQueue.copy(data, hostData, size, e).wait();
  
  for (int i = 0; i < size; ++i) {
    std::cout << "hostData["<< i << "] = " << hostData[i] << std::endl;
  }
  
  free(data, myQueue);

  return 0;
}
```
with `syclcc -O3 -std=c++17 test.cpp` (current develop branch) gives the warning:
```
In file included from test.cpp:2:
In file included from .../Documents/hipSYCL/install_dir/bin/../include/sycl/sycl.hpp:31:
In file included from .../Documents/hipSYCL/install_dir/bin/../include/sycl/../hipSYCL/sycl/sycl.hpp:55:
In file included from .../Documents/hipSYCL/install_dir/bin/../include/sycl/../hipSYCL/sycl/libkernel/accessor.hpp:38:
In file included from .../Documents/hipSYCL/install_dir/bin/../include/hipSYCL/glue/error.hpp:38:
.../Documents/hipSYCL/install_dir/bin/../include/hipSYCL/sycl/exception.hpp:68:12: warning: returning reference to local temporary object [-Wreturn-stack-address]
    return std::error_code(_error_details.info().error_code().get_code(), std::system_category());
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
```

**Expected behavior**
No reference to a temporary should be returned silencing the above warning.

**Additional context**
The C++17 Standard ([draft](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/n4713.pdf)) states in **§15.2 Temporary objects** in *(6.10)*:

> The lifetime of a temporary bound to the returned value in a function return statement (9.6.3) is not
extended; the temporary is destroyed at the end of the full-expression in the return statement.

**Comment by illuhad** [comment]
Thanks for reporting!

@sbalint98 do you want to fix this? If we really have to return a reference per the spec in that function the error code should be a class member to avoid returning a temporary.

**Comment by illuhad** [comment]
Fixed by #605
